### PR TITLE
feat: allow exporting artifacts from task and test runs

### LINKED ISSUE #1032: Allow publication of test reports
I have a java project that uses the gradle build tool. Gradle provides html test reports, which get written to `theproject/build/reports/`. That directory has the following layout:

```
tests
  |---test
           |--- classes
           |--- css
           |--- js
           |--- packages
           |--- index.html
```

It would be nice if in `garden.yml` one could somehow publish that index.html file, properly formatted, to the garden dashboard's test results.

As a workaround, currently I tweak the gradle build so it outputs informative messages to the console, in case it helps anyone trying to do the same, by adding the following to `build.gradle`:

```
test {
    testLogging {
        // set options for log level LIFECYCLE
        events "PASSED", "STARTED", "FAILED", "SKIPPED"
        exceptionFormat "short"
    }
}
```
More on that here: https://docs.gradle.org/current/dsl/org.gradle.api.tasks.testing.logging.TestLoggingContainer.html

### PR DESCRIPTION
**Pull request by edvald**
**What this PR does / why we need it**:

Users can now specify artifacts to copy after task and test runs,
for `exec`, `container` and `helm` modules.

The artifacts are stored in `.garden/artifacts` and a link to each
exported file (up to a limit of 5) is shown in the CLI after the
test/task run.

**Which issue(s) this PR fixes**:

Closes #1032 

Arguably, this also satisfies the requirements for #1253, but we should discuss that.

**Special notes for your reviewer**:

Still TODO:
- [x] Show links to artifacts in dashboard
- [x] Add section to test+task guides (waiting for #1300)

### CODE REVIEW

**Comment by edvald** [comment]
@eysi09 perhaps you could add a commit for viewing/listing the artifacts in the dashboard?

**Comment by eysi09** [comment]
Should we also include an example project with this?

File: docs/reference/module-types/container.md [hunk 1]
```diff
@@ -883,16 +870,73 @@ Example:
...
+| -------- | -------- |
+| `string` | Yes      |
+
+Example:
+
+```yaml
+tests:
+  - artifacts:
+    - source: /report/**/*
+      - source: "/output/**/*"
```
[the remark this answers is not in the document]
Comment (edvald) [reply docs/reference/module-types/container.md]: Hmm, looks like something to do with the generator.

File: garden-service/src/plugins/container/config.ts [hunk 2]
```diff
@@ -308,22 +309,9 @@ const volumeSchema = joi.object().keys({
     .description("The name of the allocated volume."),
   containerPath: joi
     .string()
-    .posixPath()
     .required()
     .description("The path where the volume should be mounted in the container."),
-  hostPath: joi
-    .string()
-    .posixPath()
-    .description(
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/plugins/container/config.ts]: This is some merge error, I'll look into that. No intention of removing this here.

File: garden-service/src/garden.ts [hunk 3]
```diff
@@ -472,8 +482,7 @@ export class Garden {
 
       if (failed.length) {
         const messages = failed.map((r) => `- ${r!.name}: ${r!.error!.message}`)
-        const names = failed.map((r) => r!.name)
-        throw new PluginError(`Failed resolving one or more providers:\n- ${names.join("\n- ")}`, {
+        throw new PluginError(`Failed resolving one or more providers:\n${messages.join("\n")}`, {
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/garden.ts]: Again, this appears to have come as part of a merge error. This PR has nothing to do with this.

File: garden-service/src/plugins/kubernetes/container/run.ts [hunk 4]
```diff
@@ -167,46 +50,23 @@ export async function runContainerService({
...
-    ignoreError: false,
-    log,
-    module,
-    namespace,
-    spec,
-    timeout: task.spec.timeout || 9999,
+    podName: `task-${module.name}-${task.name}-${Math.round(new Date().getTime())}`,
+    description: `Task '${task.name}' in container module '${module.name}'`,
+    timeout: task.spec.timeout || undefined,
+    ignoreError: true, // to ensure results get stored when an error occurs
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/plugins/kubernetes/container/run.ts]: Yes. No change in behavior there.

File: garden-service/src/plugins/kubernetes/run.ts [hunk 5]
```diff
@@ -6,112 +6,505 @@
...
-    spec.containers[0].stdin = true
-    spec.containers[0].stdinOnce = true
-    spec.containers[0].tty = true
-  }
+  const errorMetadata = { moduleName: module.name, description, args, artifacts }
+
+  if (getArtifacts) {
+    if (!command) {
+      throw new ConfigurationError(
+        deline`
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/plugins/kubernetes/run.ts]: It is. I did not choose this style.

File: garden-service/src/plugins/kubernetes/run.ts [hunk 6]
```diff
@@ -6,112 +6,505 @@
...
-  podName?: string
-  timeout?: number
-}
-
-export async function runPod({
-  provider,
-  ignoreError,
-  image,
-  interactive,
+export async function runAndCopy({
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/plugins/kubernetes/run.ts]: Yeah, thing is, I did try splitting it up but then we ended up passing pretty much all of the variables into another function. So idno, didn't feel like an improvement. It's just a fairly complex procedure.

[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/plugins/kubernetes/run.ts]: Maybe moving this onto the `PodRunner` as a method?

File: garden-service/src/plugins/kubernetes/run.ts [hunk 7]
```diff
@@ -6,112 +6,505 @@
...
+  // Prepare environment variables
+  envVars = { ...runtimeContext.envVars, ...envVars }
+  const env = uniqByName([
+    ...prepareEnvVars(envVars),
+    // If `container` is specified, include its variables as well
+    ...(container && container.env ? container.env : []),
+  ])
+
+  const getArtifacts = !interactive && artifacts && artifacts.length > 0 && artifactsPath
+  const mainContainerName = "main"
```
[the remark this answers is not in the document]
Comment (edvald) [reply garden-service/src/plugins/kubernetes/run.ts]: Not in the current implementation, no. We only use the first container atm, not the full Pod spec.

**Comment by edvald** [comment]
An example project would make sense, yes. Also still have the docs TODO.

File: garden-service/src/plugins/kubernetes/run.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (eysi09) [reply garden-service/src/plugins/kubernetes/run.ts]: > splitting it up but then we ended up passing pretty much all of the variables into another function

Yeah that's kind of what I figured. I'm fine with this being as is.

**Pull request merged**